### PR TITLE
Changing ReverseEngineering to use a better form of dependency injection

### PR DESCRIPTION
**Pull request by lajones**
See #1769.

If this looks like a good approach I'll update migrations to take the same approach.

### CODE REVIEW

File: src/EntityFramework.Commands/DatabaseTool.cs [hunk 1]
```diff
@@ -81,7 +70,12 @@ public class DatabaseTool
             var designTimeAssemblyName =
                 designTimeServicesTypeAttribute.AssemblyName ?? runtimeProviderAssemblyName;
 
-            var designTimeProvider = GetDesignTimeProvider(designTimeTypeName, designTimeAssemblyName);
+            var designTimeMetadataProviderFactory =
+                GetDesignTimeMetadataProviderFactory(designTimeTypeName, designTimeAssemblyName);
+            var serviceCollection = SetupInitialServices();
+            designTimeMetadataProviderFactory.AddMetadataProviderServices(serviceCollection);
+            designTimeMetadataProviderFactory.Create(serviceCollection);
```
Comment (bricelam) [inline src/EntityFramework.Commands/DatabaseTool.cs]: Remove this line?

File: src/EntityFramework.Relational.Design/ReverseEngineering/DesignTimeMetadataProviderFactory.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            serviceCollection.AddTransient<CSharpCodeGeneratorHelper, CSharpCodeGeneratorHelper>();
+            serviceCollection.AddTransient<ModelUtilities, ModelUtilities>();
+            serviceCollection.AddTransient<ICompilationService, RoslynCompilationService>();
+            serviceCollection.AddTransient<MetadataReferencesProvider, MetadataReferencesProvider>();
+            serviceCollection.AddTransient<ITemplating, RazorTemplating>();
+            serviceCollection.AddTransient<ReverseEngineeringGenerator, ReverseEngineeringGenerator>();
+        }
+
+        public abstract IDatabaseMetadataModelProvider Create([NotNull] ServiceCollection serviceCollection);
```
Comment (bricelam) [inline src/EntityFramework.Relational.Design/ReverseEngineering/DesignTimeMetadataProviderFactory.cs]: Why is this needed? Can't commands just get an `IDatabaseMetadataModelProvider` from the service provider?

File: src/EntityFramework.SqlServer.Design/ReverseEngineering/SqlServerDesignTimeMetadataProviderFactory.cs [hunk 3]
```diff
@@ -1,20 +1,25 @@
...
+        public override IDatabaseMetadataModelProvider Create([NotNull] ServiceCollection serviceCollection)
         {
-            Check.NotNull(serviceProvider, nameof(serviceProvider));
+            Check.NotNull(serviceCollection, nameof(serviceCollection));
 
-            return new SqlServerMetadataModelProvider(serviceProvider);
+            var serviceProvider = serviceCollection.BuildServiceProvider();
+            var logger = serviceProvider.GetRequiredService<ILogger>();
+            var modelUtilities = serviceProvider.GetRequiredService<ModelUtilities>();
+            return new SqlServerMetadataModelProvider(logger, modelUtilities);
```
Comment (bricelam) [inline src/EntityFramework.SqlServer.Design/ReverseEngineering/SqlServerDesignTimeMetadataProviderFactory.cs]: Yeah, this is silly. The thing that calls this could do the same if `SqlServerMetadataModelProvider` was registered as an `IDatabaseMetadataModelProvider`.

**Comment by bricelam** [comment]
:shipit: So much better! Assign the bug to me now. I'll port the pattern to Migrations and look into a few more ideas I've had in this area.

File: src/EntityFramework.SqlServer.Design/ReverseEngineering/SqlServerDesignTimeMetadataProviderFactory.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.SqlServer.Design/ReverseEngineering/SqlServerDesignTimeMetadataProviderFactory.cs]: Yeah - that's fine. I want to keep the majority of the service-registering in `DesignTimeMetadataProviderFactory` - so other providers can use it. But I'll update this so instead of a separate Create() method it just overrides `DesignTimeMetadataProviderFactory.AddMetadataProviderServices()`, calls the base class and then adds the service  'SqlServerMetadataModelProvider`  as an  `IDatabaseMetadataModelProvider` as you say.

**Comment by lajones** [comment]
Changes checked into release with commit a952a03, dev with commit 65cd660. Assigning to you for the Migrations changes.
